genesis/streaming.py

```diff
@@ -14,350 +14,414 @@
-		# messages, some may be duplicated if the program is killed before offsets are committed).
-		now = time.time()
-		last = self.last_commit_time if self.last_commit_time is not None else 0
-		if self._committed and (not defer or now - last > 5.):
-			print("COMMITTING", file=sys.stderr)
-			tp = [ TopicPartition(_topic, _part, _offs + 1) for (_topic, _part), _offs in self._committed.items() ]
-			#print("   TP=", tp)
-			self.c.commit(offsets=tp)
-
-			# drop all _committd offsets from _consumed; no need to commit them again if the user
-			# calls commit()
-			self._consumed = {k:v for k,v in self._consumed.items() if k not in self._committed}
-			self._committed = {}
-
-			self.last_commit_time = now
-
-	def _stream(self, filter, mapper, progress, timeout, limit):
-		import warnings
-		with warnings.catch_warnings():
-			# hide the annoying 'TqdmExperimentalWarning' warning
-			warnings.simplefilter("ignore")
-			from tqdm.autonotebook import tqdm
-
-		t = tqdm(disable=not progress, total=limit, desc='Alerts processed', unit=' alerts', mininterval=0.5, smoothing=0.5, miniters=0)
-
-		nread = 0
-		for rec, meta in self._filtered_stream(mapper=mapper, filter=filter, timeout=timeout):
-			yield (rec, meta) if self._metadata else rec
-
-			t.update()
-			nread += 1
-
-			if nread == limit:
-				break
-
-		t.close()
-
-	# returns a generator for the user-mapped messages using the filter function,
-	# possibly executed on ncores and up to maxread values
-	def __call__(self, filter=None, pool=None, progress=False, timeout=None, limit=None):
-		if pool:
-			mapper = lambda fun, vec: pool.imap(fun, vec, chunksize=100)
-		else:
-			mapper = map
-
-		yield from self._stream(filter, mapper=mapper, progress=progress, timeout=timeout, limit=limit)
-
-	def __iter__(self):
-		return self.__call__()
-
-	# producer functionality
-	def write(self, msg):
-		packet = self._serialize(msg)
-		self.p.produce(self.topics[0], packet)
-
-	def flush(self):
-		return self.p.flush()
+    c = None
+    p = None
+
+    def __init__(
+        self,
+        broker_url,
+        mode="r",
+        start_at="latest",
+        format="avro",
+        metadata=False,
+        config=None,
+    ):
+        self.groupid, self.brokers, self.topics = parse_kafka_url(broker_url)
+
+        # mode can be 'r', 'w', or 'rw'; other characters are ignored
+        assert "r" in mode or "w" in mode
+
+        if self.groupid is None:
+            # if groupid hasn't been given, emulate a low-level consumer:
+            #   - generate a random groupid
+            #   - disable committing (so we never commit the random groupid)
+            #   - start reading from the earliest (smallest) offset
+            import getpass, random
+
+            self.groupid = (
+                getpass.getuser()
+                + "-"
+                + "".join(
+                    random.choice(string.ascii_uppercase + string.digits)
+                    for _ in range(20)
```

Comment:
I must have been high(*) when I wrote this. This should be used instead:
```
"".join(random.choices(string.ascii_uppercase + string.digits, k=20))
```
(*) On a plane, w/o wifi to google the documentation.